tests/unit/lms/db/__init___test.py

```diff
@@ -0,0 +1,61 @@
+import pytest
+import sqlalchemy as sa
+
+from lms.db import BASE
+
+
+class Child(BASE):
+    __tablename__ = "child"
+    id = sa.Column(sa.Integer, primary_key=True)
+
+
+class ModelClass(BASE):
+    __tablename__ = "model_class"
+    id = sa.Column(sa.Integer, primary_key=True)
+    column = sa.Column(sa.Integer, sa.ForeignKey("child.id"))
+    relationship = sa.orm.relationship("Child")
+
+
+class TestBase:
+    def test_we_can_get_properties(self):
+        keys = {prop.key for prop in ModelClass.iter_properties()}
+
+        assert {"id", "relationship", "column"} == keys
+
+    def test_we_can_get_columns(self):
+        keys = {prop.key for prop in ModelClass.iter_columns()}
+
+        assert {"id", "column"} == keys
+
+    def test_we_can_update_from_dict(self, model):
+        model.update_from_dict(
+            {
+                "id": 4321,
+                "column": "new_value",
+                "relationship": "something",
+                "missing": "Another value",
+            }
+        )
+
+        assert model.id == 1234
+        assert model.column == "new_value"
+
+    def test_we_can_specify_keys_to_skip(self, model):
+        model.update_from_dict(
+            {"id": 4321, "column": "new_value"}, skip_keys={"column"}
+        )
+
+        assert model.id == 4321
+        assert model.column == "original_value"
+
+    def test_we_fail_to_update_when_skip_keys_is_not_a_set(self):
+        with pytest.raises(ValueError):
```

Comment:
```suggestion
        with pytest.raises(
            ValueError,
            match="Expected a set of keys to skip but found '<class 'list'>'",
        ):
```

Reply:
As a general rule, never test the string of an exception unless you are writing tests for the exception. Otherwise you get into the fun world of changing the wording of an exception message (which should by definition have no impact on the app, if it does you have other issues) and having all your tests break.

If you aren't getting enough information from the type of the exception then you probably introduce a subclass. In this case we probably should be raising a `TypeError` though: https://docs.python.org/3/library/exceptions.html#TypeError

> Passing arguments of the wrong type (e.g. passing a list when an int is expected) should result in a TypeError

I don't think you need a custom exception for every different example of I screwed up the arguments to a function though. The traceback should be enough to determine the origin, and this should really _never_ happen outside of development.

I will change this round to `TypeError` though, as that's more appropriate.

Reply:
I think it's probably a good idea to raise custom exception types (see [Modules should never raise core Python exceptions](https://utcc.utoronto.ca/~cks/space/blog/python/NeverRaiseCoreExceptions) for a good post on this) or, in cases where you don't feel it's worthwhile to create your own exception class, then do test the string. That's been my practice so far.

The problem is just that because a built-in exception class is used any mistake in the test code within the `with` or anywhere in `BaseClass.__init__()` or `BaseClass.update_from_dict()` (or anything called by those) could raise `TypeError` and cause this test to pass when it shouldn't. The test just asserts that something raised `TypeError`. It doesn't actually test that the `raise` you think was hit was in fact hit.

We have actually run into this issue a number of times before: the code has a bug, but the tests were all passing, because the test was catching the wrong exception. When it goes wrong it leads to bugs being deployed and they can be difficult to debug. 

On the other hand, what's the disadvantage of testing the string here? If you edit the string that `BaseClass` raises (seems unlikely you'd ever edit that string) then the test will start failing and you'll have to edit the string in the test as well---that doesn't seem difficult or time consuming to me, compared to potentially deploying a bug or unexpected behavior.